// @storybook/addon-viewport/dist/manager/index.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addChannel = exports.init = void 0;

var _react = _interopRequireDefault(require("react"));

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _Panel = require("./components/Panel");

var _shared = require("../shared");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var addChannel = function addChannel(api) {
  var channel = _addons.default.getChannel();

  _addons.default.addPanel(_shared.PANEL_ID, {
    title: 'Viewport',
    // eslint-disable-next-line react/prop-types
    render: function render(_ref) {
      var active = _ref.active;
      return _react.default.createElement(_Panel.Panel, {
        channel: channel,
        api: api,
        active: active
      });
    }
  });
};

exports.addChannel = addChannel;

var init = function init() {
  _addons.default.register(_shared.ADDON_ID, addChannel);
};

exports.init = init;
