// storybook-addon-source/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _addons = _interopRequireWildcard(require("@storybook/addons"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

var withSource = (0, _addons.makeDecorator)({
  name: 'withSource',
  parameterName: 'source',
  skipIfNoParametersOrOptions: true,
  allowDeprecatedUsage: true,
  wrapper: function wrapper(getStory, context, _ref) {
    var options = _ref.options,
        parameters = _ref.parameters;

    var channel = _addons["default"].getChannel();

    var storyOptions = parameters || options;

    var _ref2 = typeof storyOptions === 'string' ? {
      text: storyOptions
    } : storyOptions,
        code = _ref2.code;

    if (!code) {
      throw new Error('You must set of one of `code`  on the `source` parameter');
    }

    channel.emit('storybook/notes/add_source', code);
    return getStory(context);
  }
});
var _default = withSource;
exports["default"] = _default;
