// @storybook/addon-viewport/dist/shared/index.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_VIEWPORT = exports.INITIAL_VIEWPORTS = exports.VIEWPORT_CHANGED_EVENT_ID = exports.SET_STORY_DEFAULT_VIEWPORT_EVENT_ID = exports.CONFIGURE_VIEWPORT_EVENT_ID = exports.UPDATE_VIEWPORT_EVENT_ID = exports.PANEL_ID = exports.ADDON_ID = void 0;
var ADDON_ID = 'storybook-addon-viewport';
exports.ADDON_ID = ADDON_ID;
var PANEL_ID = "".concat(ADDON_ID, "/addon-panel");
exports.PANEL_ID = PANEL_ID;
var UPDATE_VIEWPORT_EVENT_ID = 'addon:viewport:update';
exports.UPDATE_VIEWPORT_EVENT_ID = UPDATE_VIEWPORT_EVENT_ID;
var CONFIGURE_VIEWPORT_EVENT_ID = 'addon:viewport:configure';
exports.CONFIGURE_VIEWPORT_EVENT_ID = CONFIGURE_VIEWPORT_EVENT_ID;
var SET_STORY_DEFAULT_VIEWPORT_EVENT_ID = 'addon:viewport:setStoryDefaultViewport';
exports.SET_STORY_DEFAULT_VIEWPORT_EVENT_ID = SET_STORY_DEFAULT_VIEWPORT_EVENT_ID;
var VIEWPORT_CHANGED_EVENT_ID = 'addon:viewport:viewportChanged';
exports.VIEWPORT_CHANGED_EVENT_ID = VIEWPORT_CHANGED_EVENT_ID;
var INITIAL_VIEWPORTS = {
  responsive: {
    name: 'Responsive',
    styles: {
      width: '100%',
      height: '100%',
      border: 'none',
      display: 'block',
      boxShadow: 'none'
    },
    type: 'desktop'
  },
  iphone5: {
    name: 'iPhone 5',
    styles: {
      height: '568px',
      width: '320px'
    },
    type: 'mobile'
  },
  iphone6: {
    name: 'iPhone 6',
    styles: {
      height: '667px',
      width: '375px'
    },
    type: 'mobile'
  },
  iphone6p: {
    name: 'iPhone 6 Plus',
    styles: {
      height: '736px',
      width: '414px'
    },
    type: 'mobile'
  },
  iphone8p: {
    name: 'iPhone 8 Plus',
    styles: {
      height: '736px',
      width: '414px'
    },
    type: 'mobile'
  },
  iphonex: {
    name: 'iPhone X',
    styles: {
      height: '812px',
      width: '375px'
    },
    type: 'mobile'
  },
  iphonexr: {
    name: 'iPhone XR',
    styles: {
      height: '896px',
      width: '414px'
    },
    type: 'mobile'
  },
  iphonexsmax: {
    name: 'iPhone Xs Max',
    styles: {
      height: '896px',
      width: '414px'
    },
    type: 'mobile'
  },
  ipad: {
    name: 'iPad',
    styles: {
      height: '1024px',
      width: '768px'
    },
    type: 'tablet'
  },
  ipad10p: {
    name: 'iPad Pro 10.5-in',
    styles: {
      height: '1112px',
      width: '834px'
    },
    type: 'tablet'
  },
  ipad12p: {
    name: 'iPad Pro 12.9-in',
    styles: {
      height: '1366px',
      width: '1024px'
    },
    type: 'tablet'
  },
  galaxys5: {
    name: 'Galaxy S5',
    styles: {
      height: '640px',
      width: '360px'
    },
    type: 'mobile'
  },
  galaxys9: {
    name: 'Galaxy S9',
    styles: {
      height: '1480px',
      width: '720px'
    },
    type: 'mobile'
  },
  nexus5x: {
    name: 'Nexus 5X',
    styles: {
      height: '660px',
      width: '412px'
    },
    type: 'mobile'
  },
  nexus6p: {
    name: 'Nexus 6P',
    styles: {
      height: '732px',
      width: '412px'
    },
    type: 'mobile'
  },
  pixel: {
    name: 'Pixel',
    styles: {
      height: '960px',
      width: '540px'
    },
    type: 'mobile'
  },
  pixelxl: {
    name: 'Pixel XL',
    styles: {
      height: '1280px',
      width: '720px'
    },
    type: 'mobile'
  }
};
exports.INITIAL_VIEWPORTS = INITIAL_VIEWPORTS;
var DEFAULT_VIEWPORT = 'responsive';
exports.DEFAULT_VIEWPORT = DEFAULT_VIEWPORT;
