// @storybook/addon-storysource/dist/preview.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withStorySource = withStorySource;

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _events = require("./events");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getLocation(context, locationsMap) {
  return locationsMap["".concat(context.kind, "@").concat(context.story)] || locationsMap["@".concat(context.story)];
}

function setStorySource(context, source, locationsMap) {
  var channel = _addons.default.getChannel();

  var currentLocation = getLocation(context, locationsMap);
  channel.emit(_events.EVENT_ID, {
    source: source,
    currentLocation: currentLocation,
    locationsMap: locationsMap
  });
}

function withStorySource(source) {
  var locationsMap = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return function (story, context) {
    setStorySource(context, source, locationsMap);
    return story();
  };
}
